src/main/java/org/javarosa/core/model/instance/TreeReference.java

```diff
@@ -483,14 +483,16 @@ public int hashCode() {
     }
 
     public String toString() {
-        return toString(true);
+        return toString(true, true);
     }
 
     public String toString(boolean includePredicates) {
```

Comment:
I don't think that still taking in parameters but overriding them is a good idea. Could you mark those two variants as deprecated instead?

Did you check to make sure that none of the downstream tools we know about use the toString in anything user-facing or in something that may later be machine-parsed?

Since this is a public API change, I would have expected it to be called out in the PR description.

Reply:
I think you caught this one in the process of being changed. Ultimately I removed all the overloads.

> Did you check to make sure that none of the downstream tools we know about use the toString in anything user-facing or in something that may later be machine-parsed?

Ugh, I didn't expect anyone to be coupled to the `toString()` method, and I can see some uses in Collect. It feels wrong to use `toString()` to implement business logic!

The implementation of `TreeReference.toString()` has been bugging me from the beginning. The default implementation should match the internal state of the object (with predicates and zero-based indexes), not the way around. 

In tests, we're writing and reading refs all the time, and it's annoying to have to double-check them in case you forgot to include predicates, or the multiplicity values are not what you expect.

Here's what I'll do:
- I'll revert the change in this PR
- I'll study all the places in Collect, Aggregate, Briefcase, and Validate where we use any overload of `TreeReference.toString()` and open issues to provide new API methods in JR.

Reply:
Just to illustrate how annoying the current `toString()` implementations are, take a look at some test results that I'm working on right now:

```
java.lang.AssertionError: 
Expected: is "/foo"
     but: was </foo>
Expected :is "/foo"
Actual   :</foo>
```
The difference is in a predicate, but no overload of `toString()` shows them, and I can't make my tests show them without implementing custom Hamcrest matchers, or providing a fourth new overload.